aws_lambda_powertools/event_handler/appsync.py

```diff
@@ -10,7 +11,7 @@
 
 
 class BaseRouter:
-    current_event: AppSyncResolverEventT  # type: ignore[valid-type]
+    current_event: Union[AppSyncResolverEventT, List[AppSyncResolverEventT]]  # type: ignore[valid-type]
```

Comment:
I think this requires a new specialized class: `AppSyncBatchResolver`. 

**Why**

* `current_event` will trip people using Mypy off, as it'd be difficult to know when it's a single event vs a Batch event
* Partial error handling - a customer can return individual errors (`None`) as part of its response instead of failing flat: https://docs.aws.amazon.com/appsync/latest/devguide/tutorial-lambda-resolvers.html#returning-individual-errors
    - We could argue that a customer would have to know this and do a try/except themselves, or we could also make it easier.

**Alternative**

We could introduce a new field: `current_batch`. This means always make sure it is populated with a single list item even if it's not a Batch Processor thing, so it makes it easier for customers migrating from Single Invoke to Batch Invoke (and vice-versa).

However, there's still a gap in how we can make it easier for customers wanting to fail gracefully in a Batch Invoke.

@cponfick-bhs, @thenamanpatwari - we'd love to hear your feedback to help provide an optimal UX for everyone.

Reply:
Thanks @heitorlessa  for the suggestion :-) 
I agree it might be cumbersome for the users to know what's happening. The basic idea was to do this as transparent as possible to the user and discover automatically if we deal with batch/non-batch without a need to annotate method differently. 


I like the of adding partial error handling, although I would add it as an feature in next step ( and think if we want to make it configurable and/or return user specific error). In this particular PR I would focus on making batch invoke operational as right now we're failing with the ugly exception.

The Alternative approach could work, but it seems like a breaking change if we remove current_event attribute OR additional effort to deprecate it later.

To sum up: I will pursue a design with AppSyncBatchResolver to make solution more explicit+ update the documentation, and as a next step I will add new feature to allow users return individual errors

Reply:
Absolutely agree on partial error in a separate PR - should've clarified that upfront!

oh, I wouldn't remove the `current_event`, I'd add `current_batch` **in addition** to `current_event` as a breaking a change is off the cards :) 

If you're going to work on `AppSyncBatchResolver`, we'll still need `current_event` because a customer might have a combination of Single and Batch resolvers - if that makes sense?

Reply:
I think we're onto something with `@app.batch_resolver` - allow me to summarize the call we've just had:

> **There are two fundamental problems**

1. **Correctly handling `app.current_event` when it's a Batch Invoke**. Still unsure what the value should be, whether it's a empty `AppSyncResolverEvent`, a new `AppSyncResolverEvents` to correctly identify a `list[AppSyncResolverEvent]`.
2. **Making explicit enough for the authoring experience to not rely on `app.current_event`**, since that wouldn't be correct - even if we add an Union type, this will degrade the experience for customers not using batch invoke.

What we've discussed is focusing on the second problem first by having a similar experience to the Batch Processing feature, by introducing a `@app.batch_resolver` instead of a newly specialized Class as that would confuse customers and migrating to it would be harder.

This means inverting the authoring experience for customers to only care about a single event like they do today, and we handle the Batch characteristics of it -- error handling, one event at a time, merge results into a list, partial failures, etc. 

This even opens the door to have additional tracing for every invocation so they know which field took the longest, easily with `@tracer.capture_method`.

**Example**

```python
from aws_lambda_powertools.utilities.typing import LambdaContext
from aws_lambda_powertools.utilities.data_classes import AppSyncResolverEvent

app = AppSyncResolver()

@app.batch_resolver(type_name="Query", field_name="listTodos")
def list_todos(event: AppSyncResolverEvent) -> Todo:
    return Todo() # ValueError.... [{Todo1}, None, {Todo3}]


def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

Reply:
hey @heitorlessa I applied discussed changes. Please take a look at this one more time :-) Once we get to the conclusion, I want also to modify documentation. If you can point me to the best place where we want to add it ( examples? docstrings?) that would be great.